homeassistant/components/rabbitair/fan.py

```diff
@@ -0,0 +1,148 @@
+    _power: bool | None = None
+
+    def __init__(
+        self,
+        coordinator: RabbitAirDataUpdateCoordinator,
+        entry: ConfigEntry,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__(coordinator, entry)
+
+        if self._is_model(Model.MinusA2):
+            self._attr_preset_modes = list(PRESET_MODES)
+        elif self._is_model(Model.A3):
+            # A3 does not support Pollen mode
+            self._attr_preset_modes = [
+                k for k in PRESET_MODES if k != PRESET_MODE_POLLEN
+            ]
+
+        self._attr_speed_count = len(SPEED_LIST)
+
+        self._update_state()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._update_state()
+        super()._handle_coordinator_update()
+
+    def _update_state(self) -> None:
+        data = self.coordinator.data
+
+        # Power on/off
+        self._power = data.power
+
+        # Speed as a percentage
+        if data.speed is None:
+            self._attr_percentage = None
+        elif data.speed is Speed.SuperSilent:
+            self._attr_percentage = 0
+        else:
+            self._attr_percentage = ordered_list_item_to_percentage(
+                SPEED_LIST, data.speed
+            )
+
+        # Preset mode
+        if data.mode is None:
+            self._attr_preset_mode = None
+        else:
+            # Get key by value in dictionary
+            self._attr_preset_mode = next(
+                k for k, v in PRESET_MODES.items() if v == data.mode
+            )
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set new preset mode."""
+        await self._set_state(mode=PRESET_MODES[preset_mode])
+        self._attr_preset_mode = preset_mode
+        self.async_write_ha_state()
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed of the fan, as a percentage."""
+        value = percentage_to_ordered_list_item(SPEED_LIST, percentage)
+        await self._set_state(speed=value)
+        self._attr_percentage = percentage
+        self.async_write_ha_state()
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        mode_value: Mode | None = None
+        if preset_mode is not None:
+            mode_value = PRESET_MODES[preset_mode]
+        speed_value: Speed | None = None
+        if percentage is not None:
+            speed_value = percentage_to_ordered_list_item(SPEED_LIST, percentage)
+        await self._set_state(power=True, mode=mode_value, speed=speed_value)
+        self._power = True
+        if percentage is not None:
+            self._attr_percentage = percentage
+        if preset_mode is not None:
+            self._attr_preset_mode = preset_mode
+        self.async_write_ha_state()
```

Comment:
Shouldn't it schedule an update with the coordinator instead?

Reply:
If I understand correctly, the suggestion is to request an update with `await self.coordinator.async_request_refresh()`, and update the state with real data received from the device. Yes, we can do that. But then the UI will look slow or unresponsive. The reason is that we can't get the update quick enough. It makes no sense to poll the device less than 2 seconds after sending a state change command, otherwise the state may not have changed yet.

For example, if the user turns off the device or changes the speed, the controls in the UI will show these changes with a 2-second delay. However, the advantage is that it will be the real state, not what we assumed it should be.

Also, as far as I remember, there were problems with the previous UI: if we didn't update the state fast enough, the controls jumped from the new value to the old value and back again. But it seems that now, with the new UI, there is no such effect.

In contrast, the current variant will display changes immediately, which gives the feeling that the interface is responsive to user actions and provides good feedback. However, these changes do not reflect the actual state of the device, but only our assumptions about it. The real state will be fetched in a few seconds, on the next polling cycle.

Not sure which of these options is better.